src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php

```
@@ -160,16 +176,25 @@ private function attemptSwitchUser(Request $request)
      */
     private function attemptExitUser(Request $request)
     {
-        if (false === $original = $this->getOriginalToken($this->securityContext->getToken())) {
+        $token = $this->securityContext->getToken();
+
+        // check if the authenticated user has the right role to exit switching
+        if (false === $this->accessDecisionManager->decide($token, array('ROLE_PREVIOUS_ADMIN'))) {
+            return $token;
```

I don't know here if this should throw an exception instead